### PR TITLE
feat(docker): Incremental caching for DockerHub

### PR DESCRIPTION
**Pull request by zharinov**
## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/modules/datasource/docker/index.ts [hunk 1]
```diff
@@ -883,20 +891,41 @@ export class DockerDatasource extends Datasource {
...
+    const getDockerHubTags = (): TagsResultType => {
+      const dockerHubCache = new DockerHubCache(this.http, dockerRepository);
+      return Result.wrap(dockerHubCache.getTags()).catch(getTags);
+    };
+
+    const tagsResult =
+      registryHost === 'https://index.docker.io' &&
+      process.env.RENOVATE_X_DOCKER_HUB_TAGS
+        ? getDockerHubTags()
+        : getTags();
```
Comment (rarkins) [inline lib/modules/datasource/docker/index.ts:915]: Does this still fall back to the old approach (index.docker.io) if fetching from hub.docker.com fails?

Comment (zharinov) [reply lib/modules/datasource/docker/index.ts:915]: Yes, I tried to stick to current logic as much as possible

**Review by rarkins — CHANGES_REQUESTED** [review]
I tested using `python` public image and got this WARN:

```
 WARN: Docker: Failed to parse some tags from Docker Hub (repository=renovate-reproductions/dockertags1)
       "error": {
         "message": "Schema error",
         "stack": "ZodError: Schema error\n    at Object.transform (/Users/rhys/src/github/renovatebot/renovate/lib/util/schema-utils.ts:65:21)\n    at /Users/rhys/src/github/renovatebot/renovate/node_modules/.pnpm/zod@3.22.2/node_modules/zod/lib/types.js:2870:51\n    at processTicksAndRejections (node:internal/process/task_queues:95:5)",
         "issues": {
           "68": {"digest": "Required"},
           "69": {"digest": "Required"},
           "70": {"digest": "Required"},
           "___": "... 29 more"
         }
       },
```

I think let's just silently skip any result with a missing digest, while the last_pushed should be optional.

**Comment by rarkins** [comment]
I tested with https://github.com/renovate-reproductions/dockertags1 btw

**Comment by zharinov** [comment]
One more thing is that I think we should fetch 1000 items instead of 100

File: lib/modules/datasource/docker/docker-hub-cache.ts [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+    })
+  );
+type DockerHubTag = z.infer<typeof DockerHubTag>;
+
+const DockerHubTagsPage = z
+  .object({
+    next: z.string().nullable(),
+    results: LooseArray(DockerHubTag, {
+      onError: /* istanbul ignore next */ ({ error, input: tags }) => {
+        logger.warn(
```
Comment (viceice) [inline lib/modules/datasource/docker/docker-hub-cache.ts:44]: i think debug is enough here

**Review by viceice — COMMENTED** [review]
some lint issues

**Comment by rarkins** [comment]
I believe this API doesn't return more than 100 at a time

**Comment by zharinov** [comment]
- Doesn't make much sense after #24157

**Pull request closed without merging**